### DISCUSSION TITLE
rewrite a path to variables

### DISCUSSION THREAD

**Discussion by alm494** [nginx/unit#1190] [Support]
I have a working nginx + php-fpm web application and now trying to migrate it to the new UNIT server. What is absolutely unclear in UNIT documentation is how to map URI path to the request in setting.json:

http://sample.org/app/var1/var2 to http://sample.org/app_script.php?a=var1&b=var2

Could somebody explain it with example? In regular nginx it done with 'rewrite'. Rewrite presented in unit too but I have no idea how to manage it for this simple task.

**Comment by lcrilly** [comment]
Per the [documentation](https://unit.nginx.org/configuration/#uri-rewrite) for `rewrite`

> It does not affect the [query](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4)

Which means that today, only the path segment of the URI can be rewritten.

Can you explain more about this use case?

**Reply by alm494** [reply]
It's need to build REST API service, which is in fixed standard: all queries should request with params in URI path, no query strings allowed to the end user.
So my question was how to create internally query params, not rewrite them.
This fragment seems to work and may be variables $a and $b were captured, not  sure:
`"match": {
                "uri": "~^/app/(?<a>[a-zA-Z0-9]+)/(?<b>[a-zA-Z0-9]+)$"
}`

What is the next step to pass $a and $b to PHP application as parameters? I know the workaround is rewrite php scripts to capture directly from URI but first try to avoid this work.

**Comment by hongzhidao** [comment]
Hi,
I feel there are two things here.
1. Capturing segments from the current request, such as uri, header, etc. Since Unit has matching, so it makes sense to add the feature together with it. You listed a good example.
```
"match": { "uri": "~^/app/(?<a>[a-zA-Z0-9]+)/(?<b>[a-zA-Z0-9]+)$" }
```
This feature is in our plan, we likely support it in the next version 1.33.

2. Which parts of a request can be rewritten?
Now Unit only supports URI rewrite, and we consider the arguments rewrite. But for the request headers, it's not clear yet.

About your case, there are two ways to achieve it if we need to support.
a) Make unit able to modify request headers that you can add/update/delete headers.
b) pass headers to applications like what nginx does on proxying, for example, "pass_headers".
